src/publication.rs

```diff
@@ -0,0 +1,135 @@
+        self.qos = qos;
+        self
+    }
+
+    /// Specify that this message should be [Retain::Retained].
+    pub fn retain(mut self) -> Self {
+        self.retain = Retain::Retained;
+        self
+    }
+
+    /// Specify the publication topic for this message.
+    ///
+    /// # Note
+    /// If this is called after [Publication::reply] determines a response topic, the response
+    /// topic will be overridden.
+    pub fn topic(mut self, topic: &'a str) -> Self {
+        self.topic.replace(topic);
+        self
+    }
+
+    /// Specify properties associated with this publication.
+    pub fn properties(mut self, properties: &'a [Property<'a>]) -> Self {
+        self.properties = match self.properties {
+            Properties::Slice(_) => Properties::Slice(properties),
+            Properties::CorrelatedSlice { correlation, .. } => Properties::CorrelatedSlice {
+                correlation,
+                properties,
+            },
+            _ => unimplemented!(),
+        };
+        self
+    }
+
+    /// Generate the publication as a reply to some other received message.
+    ///
+    /// # Note
+    /// The received message properties are parsed for both [Property::CorrelationData] and
+    /// [PropertyResponseTopic].
+    ///
+    /// * If correlation data is found, it is automatically appended to the
+    /// publication properties.
+    ///
+    /// * If a response topic is identified, the message topic will be
+    /// configured for it, which will override any previously-specified topic.
+    pub fn reply(mut self, received_properties: &'a Properties<'a>) -> Self {
+        if let Some(response_topic) = received_properties.into_iter().find_map(|p| {
+            if let Ok(Property::ResponseTopic(topic)) = p {
+                Some(topic.0)
+            } else {
+                None
+            }
+        }) {
+            self.topic.replace(response_topic);
+        }
+
+        // Next, copy over any correlation data to the outbound properties.
+        if let Some(correlation_data) = received_properties.into_iter().find_map(|p| {
+            if let Ok(data @ Property::CorrelationData(_)) = p {
+                Some(data)
+            } else {
+                None
+            }
+        }) {
+            self.properties = match self.properties {
+                Properties::Slice(properties) | Properties::CorrelatedSlice { properties, .. } => {
+                    Properties::CorrelatedSlice {
+                        properties,
+                        correlation: correlation_data,
+                    }
+                }
+                _ => unimplemented!(),
+            };
+        }
+
+        self
+    }
+
+    /// Generate the final publication.
+    ///
+    /// # Returns
+    /// The message to be published if a publication topic was specified. If no publication topic
+    /// was identified, an error is returned.
+    pub fn finish(self) -> Result<Pub<'a>, ProtocolError> {
```

Comment:
Can't the builder pattern be implemented directly on `Pub` (get rid of `finish()` and the `Result::Err` path)?

Reply:
Ultimately, we have to handle the case when no `topic` is identified. This should be considered an anticipated case in my opinion, since someone who only expects to `reply()` to some other message would expect that the received properties would have a `ResponseTopic`. However, the received properties may not have this topic

Reply:
Wouldn't both `Pub::new(topic) -> Self` and `Pub::reply(&self) -> Self` return structs with a valid topic?
I.e. move the payload setting (initialized as empty) into a builder method.

Reply:
No, they wouldn't. A `Pub::reply(&ReplyOptions)` still has the same problem where the `ResponseTopic` might not exist if the message you're replying to doesn't supply the property. What do you do when there's no `topic` to publish a message to? It's not something that can be sent over MQTT

Reply:
Well that would be `Pub::reply(&self) -> Option<Self>`, no?